src/compute-client/src/controller/instance.rs

```diff
@@ -882,10 +884,19 @@ where
         finishing: RowSetFinishing,
         map_filter_project: mz_expr::SafeMfpPlan,
         target_replica: Option<ReplicaId>,
+        peek_target: PeekTarget,
     ) -> Result<(), PeekError> {
-        let since = self.compute.collection(id)?.read_capabilities.frontier();
+        let since = match &peek_target {
+            PeekTarget::Index { .. } => self.compute.collection(id)?.read_capabilities.frontier(),
+            PeekTarget::Persist { .. } => self
+                .storage_controller
+                .collection(id)
+                .map_err(|_| PeekError::CollectionMissing(id))?
+                .implied_capability
```

Comment:
Fairly sure! My understanding of these is that `implied_capability` is the capability the storage controller holds as part of its standard policies (ie. the hold it would have even if nobody was currently reading the collection) and the `.read_capabilities.frontier()` is the meet of that capability and all of the in-progress readers. And while it would be legal to read as of the shared read-capability frontier - we're definitely holding the shard back that far - using the implied capability should be safe and is more conservative.

Reply:
I think that's right. There was a [Slack thread](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1700550352787449) recently about this. I think the takeaway is that compute should probably use the `implied_capability` as well in some places where it currently uses `read_capabilities`. But also the current state of things is confusing and we should probably refactor the capability handling in the controller.